### PR TITLE
[Enhancement]:Optimize bitmap index writer

### LINKED ISSUE #5328: Optimize bitmap writer performance
## Enhancement
In current `BitmapIndexWriterImpl` code,   `add_value` needs to call twice `getSizeInBytes`  the monitor the change of roaring map. The `add_value` function is called very frequently. Which have a great impact on the performance of compaction in test.

The solution to this problem is to move the calculation of the roaring bitmap size from add_value functionto thesize` function. Delay culculation.

### PR DESCRIPTION
**Pull request by xiaoyong-z**
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5328 

## Problem Summary(Required) ：

In the current BitmapIndexWriterImpl code, add_value needs to call twice getSizeInBytes the monitor the change of the roaring map. The add_value function is called very frequently. Which has a great impact on the performance of compaction in the test.
This pr optimizes the calculation of roaring size in the bitmap Index. 

The solution to this problem is to move the calculation of the roaring bitmap size from the `add_value` function` to the `size` function. Delay calculation.

### CODE REVIEW

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 1]
```diff
@@ -173,6 +171,13 @@ class BitmapIndexWriterImpl : public BitmapIndexWriter {
     uint64_t size() const override {
```
Comment (chaoyli) [inline be/src/storage/rowset/bitmap_index_writer.cpp:171]: How about the call frequency of this function?
If the frequency is low, _size_changed is non-sense.

**Review by imay — COMMENTED** [review]
I have some questions
1. is the getSizeInBytes cost dominate? or the _mem_index.find(). If all the values are different, each item will call ** getSizeInBytes** only once but the _mem_index.find() twice. if all the values are the same it will call getSizeInBytes twice but call _mem_index.find() once.
2. if the size() is called many times, getSizeInBytes will be called many times. and it won't have a better performance.

If the _mem_index.find() is the dominate, I think it is easy to eliminate it.

File: be/src/storage/rowset/bitmap_index_writer.cpp (hunk 1 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/bitmap_index_writer.cpp:171]: This function is called every we append a chunk， we need to check if the all chunk size in a segment is greater than the threshold. If it is, we need to flush the chunk.
We can omit the _size_changed here.

**Comment by xiaoyong-z** [comment]
Base compaction flame graph before optimization:
image
Base compaction flame graph after optimization:
image

**Comment by xiaoyong-z** [comment]
> I have some questions
> 
> 1. is the getSizeInBytes cost dominate? or the _mem_index.find(). If all the values are different, each item will call ** getSizeInBytes** only once but the _mem_index.find() twice. if all the values are the same it will call getSizeInBytes twice but call _mem_index.find() once.
> 2. if the size() is called many times, getSizeInBytes will be called many times. and it won't have a better performance.
> 
> If the _mem_index.find() is the dominate, I think it is easy to eliminate it.

1. In the flame graph, it shows that `getSizeInBytes` dominates.
2. i will sumbit a optimized pr later to reduce the call to minimize calls to getSizeInBytes more.

**Comment by imay** [comment]
> 1. In the flame graph, it shows that `getSizeInBytes` dominates.

It makes sense. I recheck your test case again and found that you use tiny int, which will have a lot of duplicates values. Then there will be too many getSizeInBytes calls.
So I think you can test the worst cases, which there are no same value at all. then your change may introduce more getSizeInBytes calls if size() is called many times. For example you create a bitmap index on the int64_t column, whose values are different with each other.

**Comment by xiaoyong-z** [comment]
# New experimental results
## Table
CREATE TABLE `table1` (
  `key1` int NOT NULL COMMENT "",
  `key2` datetime NOT NULL COMMENT "",
  `key3` tinyint NOT NULL COMMENT "",
  `key4` smallint NOT NULL COMMENT "",
  `key5` tinyint NOT NULL COMMENT "",
  `key6` tinyint NOT NULL COMMENT "",
  `key7` tinyint NOT NULL COMMENT "",
  `key8` tinyint NOT NULL COMMENT "",
  `key9` smallint NULL COMMENT "",
  `key10` varchar(20) NOT NULL DEFAULT "baishan" COMMENT "",
  `key11` datetime NOT NULL COMMENT "",
  `key12` datetime NOT NULL COMMENT "",
  `key13` datetime NOT NULL COMMENT "",
  `value1` bigint SUM NULL COMMENT "",
  `value2` bigint SUM NULL COMMENT "",
  `value3` bigint SUM NULL COMMENT "",
  `value4` bigint SUM NULL DEFAULT "0" COMMENT "",
  `update_time` datetime REPLACE NULL COMMENT "",
) ENGINE=OLAP
AGGREGATE KEY(`key1`, `key2`, `key3`, `key4`, `key5`, `key6`, `key7`, `key8`, `key9`, `key10`, `key11`, `key12`, `key13`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`key1`) BUCKETS 1
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
);
## Design four sets of comparison experiments
1. No bitmap index
2. bitmap index on tinyint(-128,127)
```sql
INDEX idx_bitmap_key3 (`key3`) USING BITMAP COMMENT 'key3 bitmap idx' 
```
3. bitmap index on Smallint(-32768, 32767)
```sql
INDEX idx_bitmap_key4 (`key4`) USING BITMAP COMMENT 'key4 bitmap idx'
```
4. bitmap index on Int(-2147483648, 2147483647)
```sql
INDEX idx_bitmap_key1 (`key1`) USING BITMAP COMMENT 'key1 bitmap idx'
```
## Experiment 1
base compaction time 150s
## Experiment 2
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

  | Base compaction Time | Maximum memory | Index builder time | Speed Up
-- | -- | -- | -- | --
Before Optimization | 259.837s | 1.2G | 109.837 | 1x
After Optimization | 168.927s | 1.2G | 18.027 | 6.09x

</div>
## Experiment 3
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

  | Base compaction Time | Maximum memory | Index builder time | Speed Up
-- | -- | -- | -- | --
Before Optimization | 611.759s | 1.33G | 461.759s | 1x
After Optimization | 204s | 0.31G | 54s | 8.55s

</div>
## Experiment 4
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

  | Base compaction Time | Maximum memory | Index builder time | Speed Up
-- | -- | -- | -- | --
Before Optimization | 267s | 2.53G | 117s | 1x
After Optimization | 226.449s | 0.534G | 76.449s | 1.53x

</div>

**Comment by xiaoyong-z** [comment]
# Principle of Roaring map
Following is the struct of roaring map
image
left is the key, it is the higher 16 bits of the u32_t
right is the container, there are three types of containers:
ArrayContainer, Bitmap Container, RunContainer.
image
Their relationship:
image.
Reference https://zhuanlan.zhihu.com/p/445396980
# Design idea

**Comment by xiaoyong-z** [comment]
# Design idea
When the number of keys in the roaring bitmap is relatively moderate (less than 4096, more than 512), there will be a lot of array containers with only 1 or 2 elements (while in fact each of these array containers can hold 4096 elements). Each of these containers can hold 4096 elements), creating these array containers is both very time consuming and a huge memory waste, as we saw in Experiment 3.
So my idea is to manage the elements ourselves when roaring bitmap with 1024 elements and flush them as CROARING_SERIALIZATION_ARRAY_UINT32 at flush time

**Review by imay — CHANGES_REQUESTED** [review]
If you change the data format in storage, you need to change the read logic.
But I think if you want to change the format, there are still things to consider. 
For example how to design the storage format, how to make format compatible with origin format.
So I think if you want to change the format, better to open a new issue to discuss it.
Better to do only some trivial optimization in this PR.

**Comment by xiaoyong-z** [comment]
# Experimental results After removing  hardcode roaring array
## Experiment 1
base compaction time 150s
## Experiment 2
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

  | Base compaction Time | Maximum memory | Index builder time | Speed Up
-- | -- | -- | -- | --
Before Optimization | 259.837s | 1.2G | 109.837 | 1x
After Optimization | 163.927s | 1.29G | 13.027 | 8.43x

</div>
## Experiment 3
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

  | Base compaction Time | Maximum memory | Index builder time | Speed Up
-- | -- | -- | -- | --
Before Optimization | 611.759s | 1.33G | 461.759s | 1x
After Optimization | 295.5s | 1.39G | 145.5s | 3.17x

</div>
## Experiment 4
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

  | Base compaction Time | Maximum memory | Index builder time | Speed Up
-- | -- | -- | -- | --
Before Optimization | 267s | 2.53G | 117s | 1x
After Optimization | 254s | 2.71G | 104s | 1.125x

</div>

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 2]
```diff
@@ -40,14 +40,64 @@ namespace starrocks {
...
+    uint64_t roaring_size() const { return _roaring.getSizeInBytes(false); }
+    uint64_t old_size() const { return _old_size; }
+    void set_old_size(uint64_t size) const { _old_size = size; }
+    bool size_changed() const { return _size_changed; }
+    void set_size_changed() const { _size_changed = true; }
+    void unset_size_changed() const { _size_changed = false; }
+
+    void add_rid(const rowid_t rid, uint64_t& reverted_index_size, vector<RoaringWrapper*>& size_changed_roaring_vec) {
+        _roaring.add(rid);
+        if (_enable_get_size_in_Bytes) {
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:62]:
```suggestion
        if (_enable_get_size_in_bytes) {
```

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 3]
```diff
@@ -40,14 +40,64 @@ namespace starrocks {
 
 namespace {
 
+static const size_t roaring_get_size_fastpath_threshold = 1024;
+
+class RoaringWrapper {
+public:
+    explicit RoaringWrapper(rowid_t rid)
+            : _roaring(std::move(Roaring::bitmapOf(1, rid))),
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:48]: std::move is not needed here, `Roaring::bitmapOf(1, rid)` already is a rvalue

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 4]
```diff
@@ -40,14 +40,64 @@ namespace starrocks {
 
 namespace {
 
+static const size_t roaring_get_size_fastpath_threshold = 1024;
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:43]: 1. The global variable should starts with `k_`. such as `k_roaring_get_size_fastpath_threshold`
2. It is not neccessary to make this variable a global variable, make it a class variable is enough.

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 5]
```diff
@@ -40,14 +40,64 @@ namespace starrocks {
...
+                _size_changed = true;
+                _enable_get_size_in_Bytes = true;
+                size_changed_roaring_vec.push_back(this);
+            }
+        }
+    }
+
+private:
+    Roaring _roaring;
+    mutable uint64_t _old_size;
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:88]: `_old_size` is not a good name. I think `element_count` is better.

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 6]
```diff
@@ -40,14 +40,64 @@ namespace starrocks {
...
+              _enable_get_size_in_Bytes(false){};
+    Roaring& roaring() { return _roaring; }
+    uint64_t roaring_size() const { return _roaring.getSizeInBytes(false); }
+    uint64_t old_size() const { return _old_size; }
+    void set_old_size(uint64_t size) const { _old_size = size; }
+    bool size_changed() const { return _size_changed; }
+    void set_size_changed() const { _size_changed = true; }
+    void unset_size_changed() const { _size_changed = false; }
+
+    void add_rid(const rowid_t rid, uint64_t& reverted_index_size, vector<RoaringWrapper*>& size_changed_roaring_vec) {
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:60]: If you want to change argument's value, please make it a pointer, not a reference.
Is this necessary to get estimated size each time?

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 7]
```diff
@@ -173,6 +218,13 @@ class BitmapIndexWriterImpl : public BitmapIndexWriter {
     uint64_t size() const override {
         uint64_t size = 0;
         size += _null_bitmap.getSizeInBytes(false);
+        for (RoaringWrapper* roaring_wrapper : _size_changed_roaring_vec) {
+            uint64_t new_size = roaring_wrapper->roaring_size();
+            _reverted_index_size += (new_size - roaring_wrapper->old_size());
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:223]: 1. new_size and old_size is hard to understand for me.

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 8]
```diff
@@ -173,6 +218,13 @@ class BitmapIndexWriterImpl : public BitmapIndexWriter {
     uint64_t size() const override {
         uint64_t size = 0;
         size += _null_bitmap.getSizeInBytes(false);
+        for (RoaringWrapper* roaring_wrapper : _size_changed_roaring_vec) {
+            uint64_t new_size = roaring_wrapper->roaring_size();
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:222]: You make the caller to call different function to get right size, this is not a good way.
You should make it in the RoaringWrapper do it itself.

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 9]
```diff
@@ -76,29 +126,24 @@ class BitmapIndexWriterImpl : public BitmapIndexWriter {
     void add_values(const void* values, size_t count) override {
         auto p = reinterpret_cast<const CppType*>(values);
         for (size_t i = 0; i < count; ++i) {
-            add_value(unaligned_load<CppType>(p));
+            const CppType& value = unaligned_load<CppType>(p);
+            auto it = _mem_index.find(value);
```
Comment (chaoyli) [inline be/src/storage/rowset/bitmap_index_writer.cpp:130]: 1. If the cardinality is high, does the find() dominate? It's better to test the performance at different cardinality levels.
2. We have the performance test file, you can use it and also integrate the performance test into the framework.

File: be/src/storage/rowset/bitmap_index_writer.cpp (hunk 2 above)
Comment (chaoyli) [inline be/src/storage/rowset/bitmap_index_writer.cpp:62]: When does the _enable_get_size_in_Bytes set to be true?

File: be/src/storage/rowset/bitmap_index_writer.cpp (hunk 6 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/bitmap_index_writer.cpp:60]: > Is this necessary to get estimated size each time?

For roaring whose `_element_count` less than `roaring_get_size_fastpath_threshold`, we will not push it into the `size_changed_roaring_vec`. In bitmap index's `size()`, we only update the size of roaring bitmap in `_size_changed_roaring_vec`. 
So here we need to get estimated size each time for roaring whose `_element_count` less than `roaring_get_size_fastpath_threshold`

File: be/src/storage/rowset/bitmap_index_writer.cpp (hunk 7 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/bitmap_index_writer.cpp:223]: If the `_element_count` is larger than `roaring_get_size_fastpath_threshold`, each time we will use the `roaring_wrapper->roaring_size();` to obtain the exact size of roaring bitmap.
The bitmap index's `size()` is called every time we want to append a new chunk to the segment, if the segment's size is larger than threshold, we need to first flush the segment.
So here the old_size represents the size of  roaring bitmap in the last round of chunk appended. 
The new_size represents the size of roaring bitmap in the current round of chunk appended.
The gap between them is the increasing size, and we add the diff to the _reverted_index_size.

File: be/src/storage/rowset/bitmap_index_writer.cpp (hunk 9 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/bitmap_index_writer.cpp:130]: I have tested the performance under different cardinality levels.
1. With hardcode roaring array
https://github.com/StarRocks/starrocks/pull/5329#issuecomment-1107834777
2. After removing hardcode roaring array, without replacing std::map with phmap::flat_hash_map
https://github.com/StarRocks/starrocks/pull/5329#issuecomment-1108001054

When the cardinality is high, the find search does take a bit of time. And I try to replace std::map with phmap::flat_hash_map. In experiment2 and experiment3, it has small gains(improve only 1 or 2 seconds). However, in exeriment4(the bitmap index is a random int(range from (-2147483648, 2147483647)), the version with the phmap::flat_hash_map optimization has much lower performance(around 460s, while the unoptimized version is 267s, the std::map version is 254s). In the profile, I found that phmap::flat_hash_map does frequent growing size, and this part has a lot of overhead when there is high cardinality.  So i finally still choose the std::map.

File: be/src/storage/rowset/bitmap_index_writer.cpp (hunk 2 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/bitmap_index_writer.cpp:62]: When the _element_count is less than roaring_get_size_fastpath_threshold, we will estimate the roaring bitmap size by ourselves.
When the _element_count is larger than roaring_get_size_fastpath_threshold, the `_enable_get_size_in_Bytes` is set,  and we will use the roaring bitmap's `getSizeInByte` function to obtain the exact size.

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 10]
```diff
@@ -40,14 +40,67 @@ namespace starrocks {
...
+    void update_size(uint64_t* reverted_index_size) {
+        uint64_t new_size = _roaring.getSizeInBytes(false);
+        *reverted_index_size += (new_size - _old_size);
+        _old_size = new_size;
+        _size_changed = false;
+    }
+
+private:
+    Roaring _roaring;
+    uint64_t _old_size;
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:90]: _previous_size

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 11]
```diff
@@ -40,14 +40,67 @@ namespace starrocks {
...
+
+public:
+    explicit RoaringWrapper(rowid_t rid)
+            : _roaring(Roaring::bitmapOf(1, rid)),
+              _old_size(0),
+              _element_count(1),
+              _size_changed(true),
+              _enable_get_size_in_bytes(false){};
+
+    Roaring& roaring() { return _roaring; }
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:54]:
```suggestion
    const Roaring& roaring() const { return _roaring; }
```

Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:54]:
```suggestion
    const Roaring& roaring() const { return _roaring; }
```

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 12]
```diff
@@ -40,14 +40,67 @@ namespace starrocks {
...
+                *reverted_index_size -= (roaring_get_size_fastpath_threshold * sizeof(uint32_t) + 1);
+                _size_changed = true;
+                _enable_get_size_in_bytes = true;
+                size_changed_roaring_vec->push_back(this);
+            }
+        }
+    }
+
+    void update_size(uint64_t* reverted_index_size) {
+        uint64_t new_size = _roaring.getSizeInBytes(false);
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:82]:
```suggestion
        uint64_t current_size = _roaring.getSizeInBytes(false);
```

Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:82]:
```suggestion
        uint64_t current_size = _roaring.getSizeInBytes(false);
```

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 13]
```diff
@@ -40,14 +40,67 @@ namespace starrocks {
...
+            } else {
+                *reverted_index_size -= (roaring_get_size_fastpath_threshold * sizeof(uint32_t) + 1);
+                _size_changed = true;
+                _enable_get_size_in_bytes = true;
+                size_changed_roaring_vec->push_back(this);
+            }
+        }
+    }
+
+    void update_size(uint64_t* reverted_index_size) {
```
Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:81]: I think you can get an interface like
```
size_t get_size_in_bytes() {
    if (_element_count > threshold) {
         return roaring.GetSizeInBytes();
    } else {
        return _element * 4;
    }
}
```

Comment (imay) [inline be/src/storage/rowset/bitmap_index_writer.cpp:81]: I think you can get an interface like
```
size_t get_size_in_bytes() {
    if (_element_count > threshold) {
         return roaring.GetSizeInBytes();
    } else {
        return _element * 4;
    }
}
```

File: be/src/storage/rowset/bitmap_index_writer.cpp (hunk 11 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/bitmap_index_writer.cpp:54]: this function is lator called `std::vector<Roaring*> bitmaps;`，it need a non const pointer

**Review by imay — DISMISSED** [review]

File: be/src/storage/rowset/bitmap_index_writer.cpp [hunk 14]
```diff
@@ -40,14 +40,84 @@ namespace starrocks {
...
+                _size_changed = true;
+                need_add = true;
+            }
+        } else {
+            // Add BitmapUpdateContext to _late_update_context_vector if and only
+            // it hash not been added to _late_update_context_vector before.
+            if (!_size_changed) {
+                need_add = true;
+            }
+            _size_changed = true;
```
[the remark this answers is not in the document]
Comment (xiaoyong-z) [reply be/src/storage/rowset/bitmap_index_writer.cpp:93]: 1. It depends: 
   When _disable_estimate_size is false, the need_add is equal to the _size_changed. 
   When _disable_estimate_size is true, the need_add will be true iff the size_changed is false.

**Review by chaoyli — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**